CONTRIBUTING.rst

```diff
@@ -162,3 +162,136 @@ their individual contributions.
+to the following questions:
+
+1. Is this change going to make users' lives worse?
+2. Is this change going to make the maintainers' lives worse?
+
+Code review is a collaborative process between the author and the
+reviewer to try to ensure that the answer to both of those questions
+is no.
+
+Ideally of course the change should also make one or both of the users'
+and our lives *better*, but it's OK for changes to be mostly neutral.
+The author should be presumed to have a good reason for submitting the
+change in the first place, so neutral is good enough!
+
+The rest of this section outlines specific questions reviewers should
+ask in aid of this, but the above is the most important part and where
+the guidelines below clash with the objectives above, the above wins
+out.
+
+When submitting changes it can be helpful to go through a pre-review
+pass of this yourself just to see what people will look at, but it's
+not required.
+
+~~~~~~~~~~~~~~~~~
+General Questions
+~~~~~~~~~~~~~~~~~
+
+The following are useful questions to ask for almost every change:
+
+1. Is this change too large? Could it profitably be broken up into
+   several smaller changes that could be judged on their individual
+   merits?
+2. Is the motivation for this change clear?
+3. If there is code in this change, does it make sense to at least
+   one person who isn't the author? 
+4. Is this change well covered by the review guidelines and is there
+   anything that could usefully be added to the guidelines to improve
+   that?
+
+~~~~~~~~~~~
+API Changes
+~~~~~~~~~~~
+
+Public API changes require the most careful scrutiny of all reviews,
+because they are the ones we are stuck with for the longest: Hypothesis
+follows semver and additionally tries not to break API compatibility
+wherever possible.
+
+When a public API changes we should ask the following questions:
+
+1. Is this change backwards compatible?
+2. Are there any backwards incompatible changes we'd want to make to this
+   API later and if so can we do them now instead? 
+3. How does this API handle invalid input and are the error messages clear
+   when that happens? In particular error messages should always display
+   the value that triggered the error, and ideally be specific about the
+   relevant feature of it that caused this failure (e.g. the type).
+4. How could this API otherwise be accidentally misused and is the behaviour
+   when that happens clear?
+5. Are there any cases where this API could fail silently and how can
+   those be avoided?
+6. Could this API introduce a source of flakiness? Flakiness here is
+   defined as anything where Hypothesis fails and this does not indicate
+   a bug in Hypothesis or in the way the user wrote the code or the test.
+7. Will this API be hard to support in the long-term? In particular is the
+   ability to support it very tied to the current Hypothesis internals?
+8. Is any new API adequately documented?
+9. Is the upgrade path clear for any deprecated API?
```

Comment:
Maybe see https://pyopenssl.readthedocs.io/en/latest/backward-compatibility.html as an example of guidelines for backwards incompatible changes.